src/envs/pybullet_blocks.py

```diff
@@ -0,0 +1,754 @@
+
+    def reset(self, train_or_test: str, task_idx: int) -> State:
+        state = super().reset(train_or_test, task_idx)
+        self._reset_state(state)
+        return state
+
+    def _reset_state(self, state: State) -> None:
+        """Helper for reset and testing."""
+        # Tear down the old PyBullet scene.
+        if self._held_constraint_id is not None:
+            p.removeConstraint(self._held_constraint_id,
+                               physicsClientId=self._physics_client_id)
+            self._held_constraint_id = None
+        self._held_block_id = None
+
+        # Reset robot.
+        p.resetBasePositionAndOrientation(
+            self._fetch_id,
+            self._base_pose,
+            self._base_orientation,
+            physicsClientId=self._physics_client_id)
+        rx, ry, rz, rf = state[self._robot]
+        assert np.allclose((rx, ry, rz), self._ee_home_pose)
+        joint_values = self._initial_joint_values
+        for joint_id, joint_val in zip(self._arm_joints, joint_values):
+            p.resetJointState(self._fetch_id,
+                              joint_id,
+                              joint_val,
+                              physicsClientId=self._physics_client_id)
+        for finger_id in [self._left_finger_id, self._right_finger_id]:
+            p.resetJointState(self._fetch_id,
+                              finger_id,
+                              rf,
+                              physicsClientId=self._physics_client_id)
+
+        # Reset blocks based on the state.
+        block_objs = state.get_objects(self._block_type)
+        self._block_id_to_block = {}
+        for i, block_obj in enumerate(block_objs):
+            block_id = self._block_ids[i]
+            self._block_id_to_block[block_id] = block_obj
+            bx, by, bz, _ = state[block_obj]
+            # Assume not holding in the initial state
+            assert self._get_held_block(state) is None
+            p.resetBasePositionAndOrientation(
+                block_id, [bx, by, bz],
+                self._block_orientation,
+                physicsClientId=self._physics_client_id)
+
+        # For any blocks not involved, put them out of view.
+        h = self.block_size
+        oov_x, oov_y = self._out_of_view_xy
+        for i in range(len(block_objs), len(self._block_ids)):
+            block_id = self._block_ids[i]
+            assert block_id not in self._block_id_to_block
+            p.resetBasePositionAndOrientation(
+                block_id, [oov_x, oov_y, i * h],
+                self._block_orientation,
+                physicsClientId=self._physics_client_id)
+
+        # Assert that the state was properly reconstructed.
+        reconstructed_state = self._get_state()
+        if not reconstructed_state.allclose(state):
+            print("Desired state:")
+            print(state.pretty_str())
+            print("Reconstructed state:")
+            print(reconstructed_state.pretty_str())
```

Comment:
Just a general comment but might be worth slowly migrating to logging from print in the long term - would make it easier to debug things as well

Reply:
i agree. there aren't that many print statements in our code (we try to keep them out unless they're essential), so it wouldn't take too long to do this. do you want to open an issue with some pointers regarding the best library to use @williamshen-nz ? i can work on it at some point

Reply:
Given that there aren't too many we could just use python `logging` and write some wrappers/helpers around that. 

Otherwise, there's stuff like `loguru` which can handle setups and everything in one import: https://github.com/Delgan/loguru